
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pop_finder package &#8212; pop_finder  documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="pop_finder" href="modules.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="pop-finder-package">
<h1>pop_finder package<a class="headerlink" href="#pop-finder-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-pop_finder.hp_tuning">
<span id="pop-finder-hp-tuning-module"></span><h2>pop_finder.hp_tuning module<a class="headerlink" href="#module-pop_finder.hp_tuning" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="pop_finder.hp_tuning.classifierHyperModel">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">pop_finder.hp_tuning.</span></code><code class="sig-name descname"><span class="pre">classifierHyperModel</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_shape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_classes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/hp_tuning.html#classifierHyperModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.hp_tuning.classifierHyperModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">kerastuner.engine.hypermodel.HyperModel</span></code></p>
<dl class="py method">
<dt id="pop_finder.hp_tuning.classifierHyperModel.build">
<code class="sig-name descname"><span class="pre">build</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/hp_tuning.html#classifierHyperModel.build"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.hp_tuning.classifierHyperModel.build" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a model with the specified hyperparameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>hp</strong> (<em>keras.tuners class</em>) – Class that defines how to sample hyperparameters (e.g.
RandomSearch()).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>model</strong> – Model with all the layers and specified hyperparameters.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Keras sequential model</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pop_finder.pop_finder">
<span id="pop-finder-pop-finder-module"></span><h2>pop_finder.pop_finder module<a class="headerlink" href="#module-pop_finder.pop_finder" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="pop_finder.pop_finder.get_model_name">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">get_model_name</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#get_model_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.get_model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string model name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>k</strong> (<em>int</em>) – Model number.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.run_neural_net">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">run_neural_net</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile_kfcv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infile_all</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_allele_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">patience</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_epochs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_number</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tune_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_splits</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_reps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_best_mod</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_history</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#run_neural_net"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.run_neural_net" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses input arguments from the command line to tune, train,
evaluate an ensemble of neural networks, then predicts the
population of origin for samples of unknown origin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile_kfcv</strong> (<em>string</em>) – Path to k-fold cross-validation input file, which is
a tab-delimited text file with columns x, y, pop, and
sampleID. No samples of unknown origin should be in
this file.</p></li>
<li><p><strong>infile_all</strong> (<em>string</em>) – Path to input file with all samples present (including
samples of unknown origin), which is a tab-delimited
text file with columns x, y, pop, and sampleID.</p></li>
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to VCF or hdf5 file with genetic information
for all samples (including samples of unknown origin).</p></li>
<li><p><strong>counts</strong> (<em>save_allele</em>) – Whether or not to store derived allele counts in hdf5
file (Default=False).</p></li>
<li><p><strong>save_weights</strong> (<em>boolean</em>) – Save model weights so you don’t have to retrain again
later (Default=False).</p></li>
<li><p><strong>patience</strong> (<em>int</em>) – Hyperparameter for leniency on early-stopping
(Default=10).</p></li>
<li><p><strong>batch_size</strong> (<em>int</em>) – Number of samples to use in training for each batch
(Default=32).</p></li>
<li><p><strong>max_epochs</strong> (<em>int</em>) – Number of epochs to train over.</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – Random seed (Default=None).</p></li>
<li><p><strong>train_prop</strong> (<em>float</em>) – Proportion of samples used in training (Default=0.5).</p></li>
<li><p><strong>gpu_number</strong> (<em>string</em>) – Whether to use GPUs (Default=’0’).</p></li>
<li><p><strong>tune_model</strong> (<em>boolean</em>) – Whether to tune model or just use default
(Default=False).</p></li>
<li><p><strong>n_splits</strong> (<em>int</em>) – Number of folds in k-fold cross-validation
(Default=5).</p></li>
<li><p><strong>n_reps</strong> (<em>int</em>) – Number of times to repeat k-fold cross-validation,
creating the number of models in the ensemble
(Default=5).</p></li>
<li><p><strong>save_best_mod</strong> (<em>boolean</em>) – Whether to save the model with the highest accuracy
(Default=False).</p></li>
<li><p><strong>save_dir</strong> (<em>string</em>) – Directory where results will be stored (Default=’out’).</p></li>
<li><p><strong>plot_history</strong> (<em>boolean</em>) – Whether or not to plot the training vs validation loss
over time (Default=False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>A dataframe in csv format called ‘metrics.csv’ that gives</em> – details on accuracy and performance of the best model
and overall accuracy and performance of the ensemble.</p></li>
<li><p><em>A dataframe in csv format called ‘pop_assign_freqs.csv’</em> – outlining the frequency of assignment of an individual
to each population.</p></li>
<li><p><em>A dataframe in csv format called ‘pop_assign_ensemble.csv’</em> – that specifies that top population for each individual</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pop_finder.read">
<span id="pop-finder-read-module"></span><h2>pop_finder.read module<a class="headerlink" href="#module-pop_finder.read" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="pop_finder.read.read_data">
<code class="sig-prename descclassname"><span class="pre">pop_finder.read.</span></code><code class="sig-name descname"><span class="pre">read_data</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_allele_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kfcv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/read.html#read_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.read.read_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a .zarr, .vcf, or h5py file containing genetic data and
creates subsettable data for a classifier neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>string</em>) – Path to the .zarr, .vcf, or h5py file.</p></li>
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to .txt file containing sample information
(columns are x, y, sampleID, and pop).</p></li>
<li><p><strong>save_allele_counts</strong> (<em>boolean</em>) – Saves derived allele count information (Default=False).</p></li>
<li><p><strong>kfcv</strong> (<em>boolean</em>) – If being used to test accuracy with k-fold cross-
validation (i.e. no NAs in the sample data), set to
True (Default=False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>samp_list</strong> (<em>dataframe</em>) – Contains information on corresponding sampleID and
population classifications.</p></li>
<li><p><strong>dc</strong> (<em>np.array</em>) – Array of derived allele counts.</p></li>
<li><p><strong>unknowns</strong> (<em>dataframe</em>) – If kfcv is set to False, returns a dataframe with
information about sampleID and indices for samples
of unknown origin.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pop_finder">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pop_finder" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">pop_finder</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conduct.html">Code of Conduct</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">pop_finder</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">pop_finder package</a></li>
</ul>
</li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Katie Birchard.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/source/pop_finder.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>