
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pop_finder package &#8212; pop_finder  documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="pop_finder" href="modules.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="pop-finder-package">
<h1>pop_finder package<a class="headerlink" href="#pop-finder-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-pop_finder.contour_classifier">
<span id="pop-finder-contour-classifier-module"></span><h2>pop_finder.contour_classifier module<a class="headerlink" href="#module-pop_finder.contour_classifier" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="pop_finder.contour_classifier.cont_finder">
<code class="sig-prename descclassname"><span class="pre">pop_finder.contour_classifier.</span></code><code class="sig-name descname"><span class="pre">cont_finder</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">true_dat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/contour_classifier.html#cont_finder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.contour_classifier.cont_finder" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds population in densest contour.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>true_dat</strong> (<em>pd.DataFrame</em>) – Dataframe containing x and y coordinates of all populations in
training set.</p></li>
<li><p><strong>cset</strong> (<em>matplotlib.contour.QuadContourSet</em>) – Contour values for each contour polygon.</p></li>
<li><p><strong>Returns</strong> – </p></li>
<li><p><strong>pred_pop</strong> (<em>string</em>) – Name of population in densest contour.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.contour_classifier.contour_classifier">
<code class="sig-prename descclassname"><span class="pre">pop_finder.contour_classifier.</span></code><code class="sig-name descname"><span class="pre">contour_classifier</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_contours</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_locator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen_dat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nboots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_plots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'out'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multi_iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/contour_classifier.html#contour_classifier"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.contour_classifier.contour_classifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper function that runs locator to generate a density of predictions,
then uses contour lines to choose the most likely population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_data</strong> (<em>string</em>) – Filepath to input file containing coordinates and populations of
samples, including individuals from unknown populations.</p></li>
<li><p><strong>num_contours</strong> (<em>int</em>) – Number of contours to generate and search for closest population in
(Default=10).</p></li>
<li><p><strong>run_locator</strong> (<em>boolean</em>) – Run locator and use outputs to generate classifications. If set to
False, then will look in specified save_dir for the <a href="#id1"><span class="problematic" id="id2">*</span></a>_predlocs.txt
files from a previous locator run. If set to True, ensure that
gen_dat is not None (Default=False).</p></li>
<li><p><strong>gen_dat</strong> (<em>string</em>) – Filepath to input genetic data in VCF format (Default=None).</p></li>
<li><p><strong>nboots</strong> (<em>int</em>) – Number of bootstrap iterations (Default=50).</p></li>
<li><p><strong>return_plots</strong> (<em>boolean</em>) – Return contour plots of prediction densities overlayed with true
population locations (Default=True).</p></li>
<li><p><strong>return_df</strong> (<em>boolean</em>) – If true, saves the results in a csv file in the save_dir folder
(Default=True).</p></li>
<li><p><strong>save_dir</strong> (<em>string</em>) – Folder to save results. Folder should already be in directory. If
using results from multiple iterations, only include prefix (e.g.
‘out’ for ‘out1’, ‘out2’, ‘out3’; Default=’out’).</p></li>
<li><p><strong>multi_iter</strong> (<em>int</em>) – How many times to run locator to get predictions. If the sample size
is small, it is better to run multiple iterations (Default=1).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>class_df</strong> – Dataframe containing classifications of samples of unknown origin.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.contour_classifier.kfcv">
<code class="sig-prename descclassname"><span class="pre">pop_finder.contour_classifier.</span></code><code class="sig-name descname"><span class="pre">kfcv</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen_dat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_runs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kfcv'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/contour_classifier.html#kfcv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.contour_classifier.kfcv" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs K-fold Cross-Validation and returns classification
report on model performance with given data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to tab-delimited file containing known sample
information, including x, y, pop, and sampleID columns.
NAs can be included in this file, but they will not be
used in the analysis.</p></li>
<li><p><strong>gen_dat</strong> (<em>string</em>) – Path to VCF file containing genetic information for all
samples of known origin. Do not include samples of unknown
origin in this file.</p></li>
<li><p><strong>n_splits</strong> (<em>int</em>) – Number of folds for the K-fold cross-validation (Default=5).</p></li>
<li><p><strong>n_runs</strong> (<em>int</em>) – Number of times to run K-fold cross-validation (Default=5).</p></li>
<li><p><strong>return_plot</strong> (<em>boolean</em>) – Return confusion matrix plot in output folder (Default=True).</p></li>
<li><p><strong>save_dir</strong> (<em>string</em>) – Path to output folder where confusion matrix and classification
report will be stored.</p></li>
<li><p><strong>**kwargs</strong> – Arguments for locator. See locator documentation for details.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>report</strong> – Classification report containing precision, recall, F1 score,
and accuracy for the model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pop_finder.locator_mod">
<span id="pop-finder-locator-mod-module"></span><h2>pop_finder.locator_mod module<a class="headerlink" href="#module-pop_finder.locator_mod" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="pop_finder.locator_mod.filter_snps">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">filter_snps</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_SNPs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_missing</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#filter_snps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.filter_snps" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters SNPs based on parameters specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genotypes</strong> – </p></li>
<li><p><strong>min_mac</strong> – </p></li>
<li><p><strong>max_SNPs</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>ac</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>allele counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.load_callbacks">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">load_callbacks</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">boot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jacknife</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keras_verbose</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">patience</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#load_callbacks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.load_callbacks" title="Permalink to this definition">¶</a></dt>
<dd><p>Specifies Keras callbacks, including checkpoints, early stopping,
and reducing learning rate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>boot</strong> – </p></li>
<li><p><strong>bootstrap</strong> – </p></li>
<li><p><strong>jacknife</strong> – </p></li>
<li><p><strong>out</strong> – </p></li>
<li><p><strong>keras_verbose</strong> – </p></li>
<li><p><strong>patience</strong> – </p></li>
<li><p><strong>batch_size</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>checkpointer</em></p></li>
<li><p><em>earlystop</em></p></li>
<li><p><em>reducelr</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.load_genotypes">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">load_genotypes</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gen_dat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_format</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#load_genotypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.load_genotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads genetic data and parses into relevant genotype
and sample data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gen_dat</strong> (<em>string</em>) – Path to genetic data.</p></li>
<li><p><strong>data_format</strong> (<em>string</em>) – File format of SNPs for all samples. Options include
‘vcf’, ‘zarr’, and ‘matrix’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>genotypes</strong> (<em>GenotypeArray</em>)</p></li>
<li><p><strong>samples</strong> (<em>np.array</em>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.load_network">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">load_network</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">traingen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropout_prop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nlayers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#load_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.load_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates neural network based on specifications.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>traingen</strong> – </p></li>
<li><p><strong>dropout_prop</strong> – </p></li>
<li><p><strong>nlayers</strong> – </p></li>
<li><p><strong>width</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.locator">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">locator</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen_dat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_split</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jacknife</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jacknife_prop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nboots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">patience</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_SNPs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropout_prop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nlayers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">256</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'out'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_history</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keras_verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#locator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.locator" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimates sample locations from genotype matrices. Code
modified from Battey et al. 2020:
<a class="reference external" href="https://github.com/kr-colab/locator">https://github.com/kr-colab/locator</a></p>
<dl class="simple">
<dt>sample_data<span class="classifier">string</span></dt><dd><p>Tab-delimited text file with columns ‘sampleID’, ‘x’,
and ‘y’. SampleIDs must exactly match those of the
VCF. Samples without known locations should be NA.</p>
</dd>
<dt>train_split<span class="classifier">float</span></dt><dd><p>Proportion of samples to be used for training, ranges
from 0-1 (Default=0.9).</p>
</dd>
<dt>windows<span class="classifier">boolean</span></dt><dd><p>Run windowed analysis over a single chromosome (requires
zarr input; Default=False).</p>
</dd>
<dt>window_start<span class="classifier">int</span></dt><dd><p>Starting point of windowed analysis (Default=0).</p>
</dd>
<dt>window_stop<span class="classifier">int</span></dt><dd><p>Stopping point of windowed analysis (Default=None).</p>
</dd>
<dt>window_size<span class="classifier">int</span></dt><dd><p>Size of window for windowed analysis (Default=5e5).</p>
</dd>
<dt>bootstrap<span class="classifier">boolean</span></dt><dd><p>Run bootstrap replicates by retraining on bootstrapped
data (Default=False).</p>
</dd>
<dt>jacknife<span class="classifier">boolean</span></dt><dd><p>Run jacknife uncertainty estimate on a trained network.
NOTE: we recommend this onyl as a fast heuristic –
use bootstrap option or run windowed analysis for final
results (Default=False).</p>
</dd>
<dt>nboots<span class="classifier">int</span></dt><dd><p>Number of bootstrap replicates to run (Default=50).</p>
</dd>
<dt>batch_size<span class="classifier">int</span></dt><dd><p>Number of samples per batch (Default=32).</p>
</dd>
<dt>max_epochs<span class="classifier">int</span></dt><dd><p>Number of epochs (Default=5000).</p>
</dd>
<dt>patience<span class="classifier">int</span></dt><dd><p>Number of epochs to run the optimizer after last
improvement in validation loss (Default=100).</p>
</dd>
<dt>min_mac<span class="classifier">int</span></dt><dd><p>Minimum minor allele count (Default=2).</p>
</dd>
<dt>max_SNPs<span class="classifier">int</span></dt><dd><p>Randomly select max_SNPs variants to use in the
analysis (Default=None).</p>
</dd>
<dt>impute_missing<span class="classifier">boolean</span></dt><dd><p>If False, all alleles at missing sites are ancestral
Default=False).</p>
</dd>
<dt>dropout_prop<span class="classifier">float</span></dt><dd><p>Proportion of weights to zero at the dropout layer
(Default=0.25).</p>
</dd>
<dt>nlayers<span class="classifier">int</span></dt><dd><p>Number of layers in the network (Default=10).</p>
</dd>
<dt>width<span class="classifier">int</span></dt><dd><p>Number of units per layer in the network (Default=256).</p>
</dd>
<dt>out<span class="classifier">string</span></dt><dd><p>File name stem for output (Default=’out’).</p>
</dd>
<dt>seed<span class="classifier">int</span></dt><dd><p>Random seed for train/test splits and SNP subsetting
(Default=None).</p>
</dd>
<dt>gpu_number<span class="classifier">string</span></dt><dd><p>GPU (Default=None).</p>
</dd>
<dt>plot_history<span class="classifier">boolean</span></dt><dd><p>Plot training history (Default=True).</p>
</dd>
<dt>keep_weights<span class="classifier">boolean</span></dt><dd><p>Keep model weights after training (Default=False).</p>
</dd>
<dt>load_params<span class="classifier">string</span></dt><dd><p>Path to a _params.json file to load parameters from a
previous run. Parameters from the json file will
supersede all parameters provided (Default=None).</p>
</dd>
<dt>keras_verbose<span class="classifier">int</span></dt><dd><p>Verbose argument passed to keras in model training.
0 = silent; 1 = progress bars for minibatches;
2 = show epochs. Yes, 1 is more verbose than 2. Blame
keras (Default=1).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.normalize_locs">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">normalize_locs</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#normalize_locs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.normalize_locs" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize location corrdinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>locs</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>meanlong</em></p></li>
<li><p><em>sdlong</em></p></li>
<li><p><em>meanlat</em></p></li>
<li><p><em>sdlat</em></p></li>
<li><p><em>locs</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.plot_history_fn">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">plot_history_fn</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">history</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#plot_history_fn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.plot_history_fn" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots training vs validation history over time (epochs).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>history</strong> – </p></li>
<li><p><strong>dists</strong> – </p></li>
<li><p><strong>out</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>PDF</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.predict_locs">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">predict_locs</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">predgen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sdlong</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meanlong</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sdlat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meanlat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testlocs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testgen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jacknife</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">history</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#predict_locs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.predict_locs" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict locations from trained neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – </p></li>
<li><p><strong>predgen</strong> – </p></li>
<li><p><strong>sdlong</strong> – </p></li>
<li><p><strong>meanlong</strong> – </p></li>
<li><p><strong>sdlat</strong> – </p></li>
<li><p><strong>meanlat</strong> – </p></li>
<li><p><strong>testlocs</strong> – </p></li>
<li><p><strong>pred</strong> – </p></li>
<li><p><strong>samples</strong> – </p></li>
<li><p><strong>testgen</strong> – </p></li>
<li><p><strong>verbose</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dists</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.replace_md">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">replace_md</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genotypes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#replace_md"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.replace_md" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace missing sites with values taken from binomial
distribution with parameter mean_allele_frequency.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>genotypes</strong> (<em>GenotypeArray</em>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>ac</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>allele count array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.sort_samples">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">sort_samples</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotypes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#sort_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.sort_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts samples so they are in the same order as the vcf.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>list</em>) – List of sampleIDs from the vcf file</p></li>
<li><p><strong>sample_data</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>sample_data</strong> (<em>Dataframe</em>)</p></li>
<li><p><strong>locs</strong> (<em>np.array</em>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.split_train_test">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">split_train_test</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_split</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#split_train_test"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.split_train_test" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits data into test, train, and prediction sets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ac</strong> – </p></li>
<li><p><strong>locs</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>train</em></p></li>
<li><p><em>test</em></p></li>
<li><p><em>traingen</em></p></li>
<li><p><em>testgen</em></p></li>
<li><p><em>trainlocs</em></p></li>
<li><p><em>testlocs</em></p></li>
<li><p><em>pred</em></p></li>
<li><p><em>predgen</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.locator_mod.train_network">
<code class="sig-prename descclassname"><span class="pre">pop_finder.locator_mod.</span></code><code class="sig-name descname"><span class="pre">train_network</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traingen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testgen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trainlocs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testlocs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jacknife</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checkpointer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">earlystop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reducelr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keras_verbose</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_epochs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/locator_mod.html#train_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.locator_mod.train_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Trains neural network using given parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – </p></li>
<li><p><strong>traingen</strong> – </p></li>
<li><p><strong>testgen</strong> – </p></li>
<li><p><strong>trainlocs</strong> – </p></li>
<li><p><strong>testlocs</strong> – </p></li>
<li><p><strong>bootstrap</strong> – </p></li>
<li><p><strong>jacknife</strong> – </p></li>
<li><p><strong>checkpointer</strong> – </p></li>
<li><p><strong>earlystop</strong> – </p></li>
<li><p><strong>reducelr</strong> – </p></li>
<li><p><strong>keras_verbose</strong> – </p></li>
<li><p><strong>max_epochs</strong> – </p></li>
<li><p><strong>batch_size</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>history</em></p></li>
<li><p><em>model</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pop_finder.pop_finder">
<span id="pop-finder-pop-finder-module"></span><h2>pop_finder.pop_finder module<a class="headerlink" href="#module-pop_finder.pop_finder" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="pop_finder.pop_finder.assign_plot">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">assign_plot</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensemble</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_scheme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Spectral'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#assign_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.assign_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the frequency of assignment of individuals
from unknown populations to different populations
included in the training data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>save_dir</strong> (<em>string</em>) – Path to output file where “preds.csv” lives and
also where the resulting plot will be saved.</p></li>
<li><p><strong>ensemble</strong> (<em>boolean</em>) – Set to True if multiple models used to generate assignments
(Default=False).</p></li>
<li><p><strong>col_scheme</strong> (<em>string</em>) – Colour scheme of confusion matrix. See
matplotlib.org/stable/tutorials/colors/colormaps.html
for available colour palettes (Default=”Spectral”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>assign_plot.png</strong> – PNG formatted assignment plot located in the
save_dir folder.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>PNG file</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="pop_finder.pop_finder.classifierHyperModel">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">classifierHyperModel</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_shape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_classes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#classifierHyperModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.classifierHyperModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">kerastuner.engine.hypermodel.HyperModel</span></code></p>
<dl class="py method">
<dt id="pop_finder.pop_finder.classifierHyperModel.build">
<code class="sig-name descname"><span class="pre">build</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#classifierHyperModel.build"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.classifierHyperModel.build" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a model with the specified hyperparameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>hp</strong> (<em>keras.tuners class</em>) – Class that defines how to sample hyperparameters (e.g.
RandomSearch()).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>model</strong> – Model with all the layers and specified hyperparameters.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Keras sequential model</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.hyper_tune">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">hyper_tune</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_trials</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">runs_per_trial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_prop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'out'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hyper_tune'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#hyper_tune"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.hyper_tune" title="Permalink to this definition">¶</a></dt>
<dd><p>Tunes hyperparameters of keras model for population assignment.</p>
<dl class="simple">
<dt>infile<span class="classifier">string</span></dt><dd><p>Path to VCF file containing genetic data.</p>
</dd>
<dt>sample_data<span class="classifier">string</span></dt><dd><p>Path to tab-delimited file containing columns x, y,
pop, and sampleID.</p>
</dd>
<dt>max_trials<span class="classifier">int</span></dt><dd><p>Number of trials to run for RandomSearch (Default=10).</p>
</dd>
<dt>runs_per_trial<span class="classifier">int</span></dt><dd><p>Number of runs per trial for RandomSearch (Default=10).</p>
</dd>
<dt>max_epochs<span class="classifier">int</span></dt><dd><p>Number of epochs to train model (Default=100).</p>
</dd>
<dt>train_prop<span class="classifier">float</span></dt><dd><p>Proportion of data to train on. Remaining data will be kept
as a test set and not used until final model is trained
(Default=0.8).</p>
</dd>
<dt>seed<span class="classifier">int</span></dt><dd><p>Random seed (Default=None).</p>
</dd>
<dt>save_dir<span class="classifier">string</span></dt><dd><p>Directory to save output to (Default=’out’).</p>
</dd>
<dt>mod_name<span class="classifier">string</span></dt><dd><p>Name of model in save directory (Default=’hyper_tune’).</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><strong>best_mod</strong> (<em>keras sequential model</em>) – Best model from hyperparameter tuning</p></li>
<li><p><strong>y_train</strong> (<em>pd.DataFrame</em>) – training labels</p></li>
<li><p><strong>y_val</strong> (<em>pd.DataFrame</em>) – Validation labels</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.kfcv">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">kfcv</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_splits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_reps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensemble</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kfcv_output'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_allele_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#kfcv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.kfcv" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs K-fold cross-validation to get an accuracy estimate of the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>string</em>) – Path to VCF or hdf5 file with genetic information
for all samples (including samples of unknown origin).</p></li>
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to input file with all samples present (including
samples of unknown origin), which is a tab-delimited
text file with columns x, y, pop, and sampleID.</p></li>
<li><p><strong>n_splits</strong> (<em>int</em>) – Number of folds in k-fold cross-validation
(Default=5).</p></li>
<li><p><strong>n_reps</strong> (<em>int</em>) – Number of times to repeat k-fold cross-validation,
creating the number of models in the ensemble
(Default=5).</p></li>
<li><p><strong>ensemble</strong> (<em>bool</em>) – Whether to use ensemble of models of single model (Default=False).</p></li>
<li><p><strong>save_dir</strong> (<em>string</em>) – Directory where results will be stored (Default=’kfcv_output’).</p></li>
<li><p><strong>return_plot</strong> (<em>boolean</em>) – Returns a confusion matrix of correct assignments (Default=True).</p></li>
<li><p><strong>counts</strong> (<em>save_allele</em>) – Whether or not to store derived allele counts in hdf5
file (Default=False).</p></li>
<li><p><strong>**kwargs</strong> – Keyword arguments for pop_finder function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>report</strong> (<em>pd.DataFrame</em>) – Classification report for all models.</p></li>
<li><p><strong>ensemble_report</strong> (<em>pd.DataFrame</em>) – Classification report for ensemble of models.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.pop_finder">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">pop_finder</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X_train</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_train</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unknowns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ukgen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensemble</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">try_stacking</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nbags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_prop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">predict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'out'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">patience</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_history</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#pop_finder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.pop_finder" title="Permalink to this definition">¶</a></dt>
<dd><p>Trains classifier neural network, calculates accuracy on
test set, and makes predictions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X_train</strong> (<em>np.array</em>) – Array of genetic data corresponding to train samples.</p></li>
<li><p><strong>y_train</strong> (<em>pd.DataFrame</em>) – Dataframe of train samples, including columns for samples and pops.</p></li>
<li><p><strong>X_test</strong> (<em>np.array</em>) – Array of genetic data corresponding to test samples.</p></li>
<li><p><strong>y_test</strong> (<em>pd.DataFrame</em>) – Dataframe of test samples, including columns for samples and pops.</p></li>
<li><p><strong>unknowns</strong> (<em>pd.DataFrame</em>) – Dataframe of unknowns calculated from read_data (Default=None).</p></li>
<li><p><strong>ukgen</strong> (<em>np.array</em>) – Array of genetic data corresponding to unknown samples
(Default=None).</p></li>
<li><p><strong>ensemble</strong> (<em>boolean</em>) – If set to true, will train an ensemble of models using
bootstrap aggregating (Default=False).</p></li>
<li><p><strong>try_stacking</strong> (<em>boolean</em>) – Use weights to influence ensemble model decisions. Must have
ensemble set to True to use. Use caution: with low test set sizes,
can be highly inaccurate and overfit (Default=False).</p></li>
<li><p><strong>nbags</strong> (<em>int</em>) – Number of “bags” (models) to create for the bootstrap
aggregating algorithm. This option only needs to be set if
ensemble is set to True (Default=20).</p></li>
<li><p><strong>train_prop</strong> (<em>float</em>) – Proportion of samples used in training (Default=0.8).</p></li>
<li><p><strong>mod_path</strong> (<em>string</em>) – Default=None</p></li>
<li><p><strong>predict</strong> (<em>boolean</em>) – Predict on unknown data. Must have unknowns in sample_data to use
this feature (Default=False).</p></li>
<li><p><strong>save_dir</strong> (<em>string</em>) – Directory to save results to (Default=”out”).</p></li>
<li><p><strong>save_weights</strong> (<em>boolean</em>) – Save model weights for later use (Default=False).</p></li>
<li><p><strong>patience</strong> (<em>int</em>) – How many epochs to wait before early stopping if loss has not
improved (Default=20).</p></li>
<li><p><strong>batch_size</strong> (<em>int</em>) – Default=32,</p></li>
<li><p><strong>max_epochs</strong> (<em>int</em>) – Default=100</p></li>
<li><p><strong>gpu_number</strong> (<em>string</em>) – Not in use yet, coming soon (Default=”0”).</p></li>
<li><p><strong>plot_history</strong> (<em>boolean</em>) – Plot training / validation history (Default=False).</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – Random seed for splitting data (Default=None).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>test_dict</strong> (<em>dict</em>) – Dictionary with test results.</p></li>
<li><p><strong>tot_bag_df</strong> (<em>pd.DataFrame</em>) – Dataframe with test results from ensemble.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.read_data">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">read_data</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_allele_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kfcv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#read_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.read_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a .zarr, .vcf, or h5py file containing genetic data and
creates subsettable data for a classifier neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>string</em>) – Path to the .zarr, .vcf, or h5py file.</p></li>
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to .txt file containing sample information
(columns are x, y, sampleID, and pop).</p></li>
<li><p><strong>save_allele_counts</strong> (<em>boolean</em>) – Saves derived allele count information (Default=False).</p></li>
<li><p><strong>kfcv</strong> (<em>boolean</em>) – If being used to test accuracy with k-fold cross-
validation (i.e. no NAs in the sample data), set to
True (Default=False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>samp_list</strong> (<em>dataframe</em>) – Contains information on corresponding sampleID and
population classifications.</p></li>
<li><p><strong>dc</strong> (<em>np.array</em>) – Array of derived allele counts.</p></li>
<li><p><strong>unknowns</strong> (<em>dataframe</em>) – If kfcv is set to False, returns a dataframe with
information about sampleID and indices for samples
of unknown origin.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.run_neural_net">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">run_neural_net</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_allele_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_prop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#run_neural_net"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.run_neural_net" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses input arguments from the command line to tune, train,
evaluate an ensemble of neural networks, then predicts the
population of origin for samples of unknown origin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>string</em>) – Path to VCF or hdf5 file with genetic information
for all samples (including samples of unknown origin).</p></li>
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to input file with all samples present (including
samples of unknown origin), which is a tab-delimited
text file with columns x, y, pop, and sampleID.</p></li>
<li><p><strong>counts</strong> (<em>save_allele</em>) – Whether or not to store derived allele counts in hdf5
file (Default=False).</p></li>
<li><p><strong>mod_path</strong> (<em>string</em>) – Path to tuned model. If set to None, uses default model
(Default=None).</p></li>
<li><p><strong>train_prop</strong> (<em>float</em>) – Proportion of data to be used in training the model
(Default=0.8).</p></li>
<li><p><strong>seed</strong> (<em>int</em>) – Random seed for splitting data (Default=None).</p></li>
<li><p><strong>**kwargs</strong> – Keyword arguments for pop_finder function.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.snp_rank">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">snp_rank</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'snp_rank_results'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#snp_rank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.snp_rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the most important SNPs in determining a model’s performance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>string</em>) – Path to VCF file containing genetic data.</p></li>
<li><p><strong>sample_data</strong> (<em>string</em>) – Path to tab-delimited file containing columns x, y, pop, and
sampleID.</p></li>
<li><p><strong>mod_path</strong> (<em>string</em>) – Path to tuned model. If set to None, just uses default model
(Default=None).</p></li>
<li><p><strong>save_dir</strong> (<em>string</em>) – Path to output directory (Default=”snp_rank_results”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>ranking</strong> – DataFrame containing relative importance of each SNP, where SNPs
are labelled from 1 to number of SNPs in order of appearance in
the VCF file.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="pop_finder.pop_finder.structure_plot">
<code class="sig-prename descclassname"><span class="pre">pop_finder.pop_finder.</span></code><code class="sig-name descname"><span class="pre">structure_plot</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensemble</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_scheme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Spectral'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pop_finder/pop_finder.html#structure_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pop_finder.pop_finder.structure_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes results from running the neural network with
K-fold cross-validation and creates a structure plot
showing proportion of assignment of individuals from
known populations to predicted populations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>save_dir</strong> (<em>string</em>) – Path to output file where “ensemble_preds.csv” or “preds.csv”
lives and also where the resulting plot will be saved.</p></li>
<li><p><strong>ensemble</strong> (<em>boolean</em>) – Whether ensemble of models was used to generate results or not
(Default=False).</p></li>
<li><p><strong>col_scheme</strong> (<em>string</em>) – Colour scheme of confusion matrix. See
matplotlib.org/stable/tutorials/colors/colormaps.html
for available colour palettes (Default=”Spectral”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>structure_plot.png</strong> – PNG formatted structure plot located in the
save_dir folder.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>PNG file</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pop_finder">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pop_finder" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">pop_finder</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">pop_finder</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">pop_finder package</a></li>
</ul>
</li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Katie Birchard.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.5.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/source/pop_finder.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>